File: src/joypad.rs
use std::sync::Arc;
use std::thread;

use gilrs::{Axis, Button, EventType, Gilrs};
use slint::ComponentHandle;
use slint::platform::{Key, WindowEvent};

use crate::terminal::TerminalManager;
use crate::MainWindow;

const STICK_THRESHOLD: f32 = 0.5;

#[derive(Clone, Copy, PartialEq)]
enum Dir {
    None,
    Left,
    Right,
    Up,
    Down,
}

pub fn register_joypad(
    weak_window: slint::Weak<MainWindow>,
    terminal: Arc<TerminalManager>,
) {
    thread::spawn(move || {
        let mut gilrs = match Gilrs::new() {
            Ok(g) => g,
            Err(e) => {
                eprintln!("gilrs init error: {e}");
                return;
            }
        };

        let mut stick_x = Dir::None;
        let mut stick_y = Dir::None;
        let mut l2 = false;
        let mut r2 = false;

        while let Some(event) = gilrs.next_event() {
            match event.event {
                EventType::ButtonPressed(button, _) => match button {
                    Button::LeftTrigger2 => l2 = true,
                    Button::RightTrigger2 => r2 = true,
                    _ => {
                        if let Some(key) = button_to_key(button) {
                            send_key(&weak_window, key, true);
                        }
                    }
                },
                EventType::ButtonReleased(button, _) => match button {
                    Button::LeftTrigger2 => l2 = false,
                    Button::RightTrigger2 => r2 = false,
                    _ => {
                        if let Some(key) = button_to_key(button) {
                            send_key(&weak_window, key, false);
                        }
                    }
                },
                EventType::AxisChanged(axis, value, _) => match axis {
                    Axis::LeftStickX => {
                        let new_dir = dir_from_x(value);
                        handle_stick(&weak_window, &mut stick_x, new_dir);
                    }
                    Axis::LeftStickY => {
                        let new_dir = dir_from_y(value);
                        handle_stick(&weak_window, &mut stick_y, new_dir);
                    }
                    _ => {}
                },
                _ => {}
            }

            // L2+R2 together terminates a running interactive program and
            // returns to the settings list.
            if l2 && r2 && terminal.is_running() {
                terminal.kill(&weak_window);
                l2 = false;
                r2 = false;
            }
        }
    });
}

fn dir_from_x(value: f32) -> Dir {
    if value < -STICK_THRESHOLD {
        Dir::Left
    } else if value > STICK_THRESHOLD {
        Dir::Right
    } else {
        Dir::None
    }
}

fn dir_from_y(value: f32) -> Dir {
    if value < -STICK_THRESHOLD {
        Dir::Up
    } else if value > STICK_THRESHOLD {
        Dir::Down
    } else {
        Dir::None
    }
}

fn handle_stick(weak_window: &slint::Weak<MainWindow>, current: &mut Dir, new: Dir) {
    if *current == new {
        return;
    }

    let (release, press) = match (*current, new) {
        (Dir::Left, Dir::None) => (Some(Key::LeftArrow), None),
        (Dir::Right, Dir::None) => (Some(Key::RightArrow), None),
        (Dir::Up, Dir::None) => (Some(Key::UpArrow), None),
        (Dir::Down, Dir::None) => (Some(Key::DownArrow), None),
        (Dir::Left, Dir::Right) => (Some(Key::LeftArrow), Some(Key::RightArrow)),
        (Dir::Right, Dir::Left) => (Some(Key::RightArrow), Some(Key::LeftArrow)),
        (Dir::Up, Dir::Down) => (Some(Key::UpArrow), Some(Key::DownArrow)),
        (Dir::Down, Dir::Up) => (Some(Key::DownArrow), Some(Key::UpArrow)),
        (Dir::None, dir) => (
            None,
            match dir {
                Dir::Left => Some(Key::LeftArrow),
                Dir::Right => Some(Key::RightArrow),
                Dir::Up => Some(Key::UpArrow),
                Dir::Down => Some(Key::DownArrow),
                _ => None,
            },
        ),
        _ => (None, None),
    };

    if let Some(key) = release {
        send_key(weak_window, key, false);
    }
    if let Some(key) = press {
        send_key(weak_window, key, true);
    }

    *current = new;
}

fn button_to_key(button: Button) -> Option<Key> {
    match button {
        Button::DPadUp => Some(Key::UpArrow),
        Button::DPadDown => Some(Key::DownArrow),
        Button::DPadLeft => Some(Key::LeftArrow),
        Button::DPadRight => Some(Key::RightArrow),
        Button::South => Some(Key::Return),
        Button::East => Some(Key::Backspace),
        Button::Start => Some(Key::Space),
        Button::Select => Some(Key::Menu),
        _ => None,
    }
}

fn send_key(weak_window: &slint::Weak<MainWindow>, key: Key, pressed: bool) {
    let weak_window = weak_window.clone();
    slint::invoke_from_event_loop(move || {
        if let Some(window) = weak_window.upgrade() {
            let event = if pressed {
                WindowEvent::KeyPressed { text: key.into() }
            } else {
                WindowEvent::KeyReleased { text: key.into() }
            };
            window.window().dispatch_event(event);
        }
    })
    .ok();
}

File: src/joypad.cpp
#include "joypad.h"

#include <array>
#include <print>
#include <string>

#include <fcntl.h>
#include <linux/input.h>
#include <poll.h>
#include <unistd.h>

#include "unique_fd.h"

constexpr float STICK_THRESHOLD = 0.5f;

enum struct Dir { None, Left, Right, Up, Down };

static Dir dir_from_x(float value) {
    if (value < -STICK_THRESHOLD) return Dir::Left;
    if (value > STICK_THRESHOLD) return Dir::Right;
    return Dir::None;
}

static Dir dir_from_y(float value) {
    if (value < -STICK_THRESHOLD) return Dir::Up;
    if (value > STICK_THRESHOLD) return Dir::Down;
    return Dir::None;
}

static std::string_view key_name_for_dir(Dir d) {
    switch (d) {
        case Dir::Left:  return "ArrowLeft";
        case Dir::Right: return "ArrowRight";
        case Dir::Up:    return "ArrowUp";
        case Dir::Down:  return "ArrowDown";
        default:         return {};
    }
}

static void send_key_press(slint::ComponentWeakHandle<MainWindow> weak_window, std::string_view key) {
    std::string key_str(key);
    slint::invoke_from_event_loop([weak_window, key_str]() {
        if (auto w = weak_window.lock()) {
            (*w)->window().dispatch_key_press_event(slint::SharedString(key_str));
        }
    });
}

static void send_key_release(slint::ComponentWeakHandle<MainWindow> weak_window, std::string_view key) {
    std::string key_str(key);
    slint::invoke_from_event_loop([weak_window, key_str]() {
        if (auto w = weak_window.lock()) {
            (*w)->window().dispatch_key_release_event(slint::SharedString(key_str));
        }
    });
}

static void send_key(slint::ComponentWeakHandle<MainWindow> weak_window, std::string_view key, bool pressed) {
    if (pressed) {
        send_key_press(weak_window, key);
    } else {
        send_key_release(weak_window, key);
    }
}

static std::string_view button_to_key(uint16_t code) {
    switch (code) {
        case BTN_DPAD_UP:    return "ArrowUp";
        case BTN_DPAD_DOWN:  return "ArrowDown";
        case BTN_DPAD_LEFT:  return "ArrowLeft";
        case BTN_DPAD_RIGHT: return "ArrowRight";
        case BTN_SOUTH:      return "Return";
        case BTN_EAST:       return "Backspace";
        case BTN_START:      return "Space";
        case BTN_SELECT:     return "Menu";
        default:             return {};
    }
}

static void handle_stick(slint::ComponentWeakHandle<MainWindow> weak_window,
                         Dir& current, Dir new_dir) {
    if (current == new_dir) return;

    Dir release_dir = Dir::None;
    Dir press_dir = Dir::None;

    if (new_dir == Dir::None) {
        release_dir = current;
    } else if (current == Dir::None) {
        press_dir = new_dir;
    } else {
        release_dir = current;
        press_dir = new_dir;
    }

    if (release_dir != Dir::None) {
        auto name = key_name_for_dir(release_dir);
        if (!name.empty()) send_key(weak_window, name, false);
    }
    if (press_dir != Dir::None) {
        auto name = key_name_for_dir(press_dir);
        if (!name.empty()) send_key(weak_window, name, true);
    }

    current = new_dir;
}

JoypadManager::~JoypadManager() {
    stop();
}

void JoypadManager::stop() {
    thread_.request_stop();
}

void JoypadManager::start(slint::ComponentWeakHandle<MainWindow> weak_window,
                           std::shared_ptr<TerminalManager> terminal) {
    thread_ = std::jthread([weak_window, terminal](std::stop_token stop_token) {
        UniqueFd gamepad_fd;

        for (int i = 0; i < 32 && !stop_token.stop_requested(); i++) {
            std::string path = "/dev/input/event" + std::to_string(i);
            int raw_fd = open(path.c_str(), O_RDONLY | O_NONBLOCK);
            if (raw_fd < 0) continue;

            std::array<char, 256> name{};
            ioctl(raw_fd, EVIOCGNAME(name.size()), name.data());

            unsigned long bits[KEY_MAX / 8 / sizeof(unsigned long) + 1] = {};
            if (ioctl(raw_fd, EVIOCGBIT(EV_KEY, sizeof(bits)), bits) >= 0) {
                if (bits[BTN_DPAD_UP / 8 / sizeof(unsigned long)] &
                    (1UL << ((BTN_DPAD_UP / 8) % sizeof(unsigned long)))) {
                    gamepad_fd.reset(raw_fd);
                    break;
                }
            }
            ::close(raw_fd);
        }

        if (!gamepad_fd.valid()) {
            std::println(std::cerr, "joypad: no gamepad found");
            return;
        }

        std::println(std::cerr, "joypad: device opened");

        Dir stick_x = Dir::None;
        Dir stick_y = Dir::None;
        bool l2 = false;
        bool r2 = false;

        struct pollfd pfd = {};
        pfd.fd = gamepad_fd.get();
        pfd.events = POLLIN;

        struct input_event ev;
        while (!stop_token.stop_requested()) {
            int ret = poll(&pfd, 1, 100);
            if (ret <= 0) continue;
            if (!(pfd.revents & POLLIN)) continue;

            ssize_t n = ::read(gamepad_fd.get(), &ev, sizeof(ev));
            if (n != sizeof(ev)) break;

            if (ev.type == EV_KEY) {
                if (ev.code == BTN_TL2) {
                    l2 = (ev.value != 0);
                } else if (ev.code == BTN_TR2) {
                    r2 = (ev.value != 0);
                } else {
                    auto key = button_to_key(ev.code);
                    if (!key.empty()) {
                        send_key(weak_window, key, ev.value != 0);
                    }
                }
            } else if (ev.type == EV_ABS) {
                if (ev.code == ABS_X) {
                    float value = ev.value / 32767.0f;
                    Dir new_dir = dir_from_x(value);
                    handle_stick(weak_window, stick_x, new_dir);
                } else if (ev.code == ABS_Y) {
                    float value = ev.value / 32767.0f;
                    Dir new_dir = dir_from_y(value);
                    handle_stick(weak_window, stick_y, new_dir);
                }
            }

            if (l2 && r2 && terminal->is_running()) {
                terminal->terminate(weak_window);
                l2 = false;
                r2 = false;
            }
        }
    });
}
